File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/go/pkg/verification/verification_test.go
package verification

import (
	"testing"

	"github.com/ThirdKeyAi/schemapin/go/pkg/bundle"
	"github.com/ThirdKeyAi/schemapin/go/pkg/core"
	gocrypto "github.com/ThirdKeyAi/schemapin/go/pkg/crypto"
	"github.com/ThirdKeyAi/schemapin/go/pkg/discovery"
	"github.com/ThirdKeyAi/schemapin/go/pkg/resolver"
	"github.com/ThirdKeyAi/schemapin/go/pkg/revocation"
)

func makeKeyAndSign(schema map[string]interface{}) (string, string, string) {
	km := gocrypto.NewKeyManager()
	sm := gocrypto.NewSignatureManager()
	c := core.NewSchemaPinCore()

	privKey, err := km.GenerateKeypair()
	if err != nil {
		panic(err)
	}

	pubPEM, err := km.ExportPublicKeyPEM(&privKey.PublicKey)
	if err != nil {
		panic(err)
	}

	schemaHash, err := c.CanonicalizeAndHash(schema)
	if err != nil {
		panic(err)
	}

	sig, err := sm.SignSchemaHash(schemaHash, privKey)
	if err != nil {
		panic(err)
	}

	fp, err := km.CalculateKeyFingerprintFromPEM(pubPEM)
	if err != nil {
		panic(err)
	}

	return pubPEM, sig, fp
}

func TestKeyPinStoreFirstUse(t *testing.T) {
	store := NewKeyPinStore()
	result := store.CheckAndPin("tool1", "example.com", "sha256:aaa")
	if result != PinFirstUse {
		t.Errorf("expected first_use, got %s", result)
	}
}

func TestKeyPinStorePinned(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("tool1", "example.com", "sha256:aaa")
	result := store.CheckAndPin("tool1", "example.com", "sha256:aaa")
	if result != PinPinned {
		t.Errorf("expected pinned, got %s", result)
	}
}

func TestKeyPinStoreChanged(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("tool1", "example.com", "sha256:aaa")
	result := store.CheckAndPin("tool1", "example.com", "sha256:bbb")
	if result != PinChanged {
		t.Errorf("expected changed, got %s", result)
	}
}

func TestKeyPinStoreDifferentTools(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("tool1", "example.com", "sha256:aaa")
	result := store.CheckAndPin("tool2", "example.com", "sha256:bbb")
	if result != PinFirstUse {
		t.Errorf("expected first_use, got %s", result)
	}
}

func TestKeyPinStoreDifferentDomains(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("tool1", "a.com", "sha256:aaa")
	result := store.CheckAndPin("tool1", "b.com", "sha256:bbb")
	if result != PinFirstUse {
		t.Errorf("expected first_use, got %s", result)
	}
}

func TestKeyPinStoreJSONRoundtrip(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("tool1", "example.com", "sha256:aaa")
	store.CheckAndPin("tool2", "other.com", "sha256:bbb")

	jsonStr, err := store.ToJSON()
	if err != nil {
		t.Fatalf("failed to serialize: %v", err)
	}

	restored, err := FromJSON(jsonStr)
	if err != nil {
		t.Fatalf("failed to deserialize: %v", err)
	}

	if restored.CheckAndPin("tool1", "example.com", "sha256:aaa") != PinPinned {
		t.Error("expected pinned for tool1")
	}
	if restored.CheckAndPin("tool2", "other.com", "sha256:bbb") != PinPinned {
		t.Error("expected pinned for tool2")
	}
}

func TestKeyPinStoreGetPinned(t *testing.T) {
	store := NewKeyPinStore()
	store.CheckAndPin("tool1", "example.com", "sha256:aaa")
	if store.GetPinned("tool1", "example.com") != "sha256:aaa" {
		t.Error("expected sha256:aaa")
	}
	if store.GetPinned("tool2", "example.com") != "" {
		t.Error("expected empty string for unknown tool")
	}
}

func TestVerifySchemaOfflineHappyPath(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM, sig, _ := makeKeyAndSign(schema)

	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Test Dev",
		PublicKeyPEM:  pubPEM,
	}
	store := NewKeyPinStore()
	result := VerifySchemaOffline(schema, sig, "example.com", "tool1", disc, nil, store)

	if !result.Valid {
		t.Errorf("expected valid, got error: %s", result.ErrorMessage)
	}
	if result.Domain != "example.com" {
		t.Errorf("expected domain example.com, got %s", result.Domain)
	}
	if result.DeveloperName != "Test Dev" {
		t.Errorf("expected developer Test Dev, got %s", result.DeveloperName)
	}
	if result.KeyPinning == nil || result.KeyPinning.Status != "first_use" {
		t.Error("expected key_pinning.status = first_use")
	}
}

func TestVerifySchemaOfflinePinnedOnSecondCall(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM, sig, _ := makeKeyAndSign(schema)

	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Test Dev",
		PublicKeyPEM:  pubPEM,
	}
	store := NewKeyPinStore()
	VerifySchemaOffline(schema, sig, "example.com", "tool1", disc, nil, store)
	result := VerifySchemaOffline(schema, sig, "example.com", "tool1", disc, nil, store)

	if !result.Valid {
		t.Error("expected valid")
	}
	if result.KeyPinning.Status != "pinned" {
		t.Error("expected pinned status")
	}
}

func TestVerifySchemaOfflineInvalidSignature(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM, _, _ := makeKeyAndSign(schema)

	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Test Dev",
		PublicKeyPEM:  pubPEM,
	}
	store := NewKeyPinStore()
	result := VerifySchemaOffline(schema, "invalid_sig", "example.com", "tool1", disc, nil, store)

	if result.Valid {
		t.Error("expected invalid")
	}
	if result.ErrorCode != ErrSignatureInvalid {
		t.Errorf("expected signature_invalid, got %s", result.ErrorCode)
	}
}

func TestVerifySchemaOfflineTamperedSchema(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM, sig, _ := makeKeyAndSign(schema)

	tampered := map[string]interface{}{"name": "test_tool", "description": "TAMPERED"}
	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Test Dev",
		PublicKeyPEM:  pubPEM,
	}
	store := NewKeyPinStore()
	result := VerifySchemaOffline(tampered, sig, "example.com", "tool1", disc, nil, store)

	if result.Valid {
		t.Error("expected invalid")
	}
	if result.ErrorCode != ErrSignatureInvalid {
		t.Errorf("expected signature_invalid, got %s", result.ErrorCode)
	}
}

func TestVerifySchemaOfflineRevokedKeySimpleList(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM, sig, fp := makeKeyAndSign(schema)

	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Test Dev",
		PublicKeyPEM:  pubPEM,
		RevokedKeys:   []string{fp},
	}
	store := NewKeyPinStore()
	result := VerifySchemaOffline(schema, sig, "example.com", "tool1", disc, nil, store)

	if result.Valid {
		t.Error("expected invalid")
	}
	if result.ErrorCode != ErrKeyRevoked {
		t.Errorf("expected key_revoked, got %s", result.ErrorCode)
	}
}

func TestVerifySchemaOfflineRevokedKeyStandaloneDoc(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM, sig, fp := makeKeyAndSign(schema)

	disc := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Test Dev",
		PublicKeyPEM:  pubPEM,
	}
	rev := revocation.BuildRevocationDocument("example.com")
	revocation.AddRevokedKey(rev, fp, revocation.ReasonKeyCompromise)

	store := NewKeyPinStore()
	result := VerifySchemaOffline(schema, sig, "example.com", "tool1", disc, rev, store)

	if result.Valid {
		t.Error("expected invalid")
	}
	if result.ErrorCode != ErrKeyRevoked {
		t.Errorf("expected key_revoked, got %s", result.ErrorCode)
	}
}

func TestVerifySchemaOfflineKeyPinChangeRejected(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM1, sig1, _ := makeKeyAndSign(schema)
	pubPEM2, sig2, _ := makeKeyAndSign(schema)

	disc1 := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Dev",
		PublicKeyPEM:  pubPEM1,
	}
	disc2 := &discovery.WellKnownResponse{
		SchemaVersion: "1.2",
		DeveloperName: "Dev",
		PublicKeyPEM:  pubPEM2,
	}

	store := NewKeyPinStore()
	r1 := VerifySchemaOffline(schema, sig1, "example.com", "tool1", disc1, nil, store)
	if !r1.Valid {
		t.Fatalf("first verification failed: %s", r1.ErrorMessage)
	}

	r2 := VerifySchemaOffline(schema, sig2, "example.com", "tool1", disc2, nil, store)
	if r2.Valid {
		t.Error("expected invalid for key change")
	}
	if r2.ErrorCode != ErrKeyPinMismatch {
		t.Errorf("expected key_pin_mismatch, got %s", r2.ErrorCode)
	}
}

func TestVerifySchemaOfflineInvalidDiscovery(t *testing.T) {
	store := NewKeyPinStore()
	result := VerifySchemaOffline(
		map[string]interface{}{"name": "test"},
		"sig", "example.com", "tool1",
		&discovery.WellKnownResponse{SchemaVersion: "1.2"},
		nil, store,
	)

	if result.Valid {
		t.Error("expected invalid")
	}
	if result.ErrorCode != ErrDiscoveryInvalid {
		t.Errorf("expected discovery_invalid, got %s", result.ErrorCode)
	}
}

func TestVerifySchemaOfflineNilDiscovery(t *testing.T) {
	store := NewKeyPinStore()
	result := VerifySchemaOffline(
		map[string]interface{}{"name": "test"},
		"sig", "example.com", "tool1",
		nil, nil, store,
	)

	if result.Valid {
		t.Error("expected invalid")
	}
	if result.ErrorCode != ErrDiscoveryInvalid {
		t.Errorf("expected discovery_invalid, got %s", result.ErrorCode)
	}
}

func TestVerifySchemaWithResolverHappyPath(t *testing.T) {
	schema := map[string]interface{}{"name": "test_tool", "description": "A test"}
	pubPEM, sig, _ := makeKeyAndSign(schema)

	b := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	b.Documents = append(b.Documents, bundle.BundledDiscovery{
		Domain: "example.com",
		WellKnown: discovery.WellKnownResponse{
			SchemaVersion: "1.2",
			DeveloperName: "Bundle Dev",
			PublicKeyPEM:  pubPEM,
		},
	})

	r := resolver.NewTrustBundleResolver(b)
	store := NewKeyPinStore()

	result := VerifySchemaWithResolver(schema, sig, "example.com", "tool1", r, store)
	if !result.Valid {
		t.Errorf("expected valid, got error: %s", result.ErrorMessage)
	}
	if result.DeveloperName != "Bundle Dev" {
		t.Errorf("expected Bundle Dev, got %s", result.DeveloperName)
	}
}

func TestVerifySchemaWithResolverMissingDomain(t *testing.T) {
	b := bundle.NewTrustBundle("2026-01-01T00:00:00Z")
	r := resolver.NewTrustBundleResolver(b)
	store := NewKeyPinStore()

	result := VerifySchemaWithResolver(
		map[string]interface{}{"name": "test"},
		"sig", "missing.com", "tool1", r, store,
	)

	if result.Valid {
		t.Error("expected invalid")
	}
	if result.ErrorCode != ErrDiscoveryFetchFailed {
		t.Errorf("expected discovery_fetch_failed, got %s", result.ErrorCode)
	}
}

File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/javascript/tests/verification.test.js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';

import { KeyManager, SignatureManager } from '../src/crypto.js';
import { SchemaPinCore } from '../src/core.js';
import { createTrustBundle, createBundledDiscovery } from '../src/bundle.js';
import { TrustBundleResolver } from '../src/resolver.js';
import {
    buildRevocationDocument,
    addRevokedKey,
    RevocationReason
} from '../src/revocation.js';
import {
    ErrorCode,
    KeyPinStore,
    verifySchemaOffline,
    verifySchemaWithResolver
} from '../src/verification.js';

function makeKeyAndSign(schema) {
    const { privateKey, publicKey } = KeyManager.generateKeypair();
    const pubPem = KeyManager.exportPublicKeyPem(publicKey);
    const schemaHash = SchemaPinCore.canonicalizeAndHash(schema);
    const sig = SignatureManager.signSchemaHash(schemaHash, privateKey);
    const fp = KeyManager.calculateKeyFingerprint(pubPem);
    return { pubPem, sig, fp };
}

describe('KeyPinStore', () => {
    it('should return first_use for new tool@domain', () => {
        const store = new KeyPinStore();
        assert.equal(store.checkAndPin('tool1', 'example.com', 'sha256:aaa'), 'first_use');
    });

    it('should return pinned for same fingerprint', () => {
        const store = new KeyPinStore();
        store.checkAndPin('tool1', 'example.com', 'sha256:aaa');
        assert.equal(store.checkAndPin('tool1', 'example.com', 'sha256:aaa'), 'pinned');
    });

    it('should return changed for different fingerprint', () => {
        const store = new KeyPinStore();
        store.checkAndPin('tool1', 'example.com', 'sha256:aaa');
        assert.equal(store.checkAndPin('tool1', 'example.com', 'sha256:bbb'), 'changed');
    });

    it('should keep different tools independent', () => {
        const store = new KeyPinStore();
        store.checkAndPin('tool1', 'example.com', 'sha256:aaa');
        assert.equal(store.checkAndPin('tool2', 'example.com', 'sha256:bbb'), 'first_use');
    });

    it('should keep different domains independent', () => {
        const store = new KeyPinStore();
        store.checkAndPin('tool1', 'a.com', 'sha256:aaa');
        assert.equal(store.checkAndPin('tool1', 'b.com', 'sha256:bbb'), 'first_use');
    });

    it('should roundtrip through JSON', () => {
        const store = new KeyPinStore();
        store.checkAndPin('tool1', 'example.com', 'sha256:aaa');
        store.checkAndPin('tool2', 'other.com', 'sha256:bbb');

        const json = store.toJSON();
        const restored = KeyPinStore.fromJSON(json);

        assert.equal(restored.checkAndPin('tool1', 'example.com', 'sha256:aaa'), 'pinned');
        assert.equal(restored.checkAndPin('tool2', 'other.com', 'sha256:bbb'), 'pinned');
    });

    it('should get pinned fingerprint', () => {
        const store = new KeyPinStore();
        store.checkAndPin('tool1', 'example.com', 'sha256:aaa');
        assert.equal(store.getPinned('tool1', 'example.com'), 'sha256:aaa');
        assert.equal(store.getPinned('tool2', 'example.com'), null);
    });
});

describe('verifySchemaOffline', () => {
    it('should pass for valid schema and signature', () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const { pubPem, sig } = makeKeyAndSign(schema);

        const discovery = {
            schema_version: '1.2',
            developer_name: 'Test Dev',
            public_key_pem: pubPem
        };
        const store = new KeyPinStore();
        const result = verifySchemaOffline(schema, sig, 'example.com', 'tool1', discovery, null, store);

        assert.equal(result.valid, true);
        assert.equal(result.domain, 'example.com');
        assert.equal(result.developer_name, 'Test Dev');
        assert.equal(result.key_pinning.status, 'first_use');
        assert.equal(result.error_code, undefined);
    });

    it('should return pinned on second call', () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const { pubPem, sig } = makeKeyAndSign(schema);

        const discovery = {
            schema_version: '1.2',
            developer_name: 'Test Dev',
            public_key_pem: pubPem
        };
        const store = new KeyPinStore();
        verifySchemaOffline(schema, sig, 'example.com', 'tool1', discovery, null, store);
        const result = verifySchemaOffline(schema, sig, 'example.com', 'tool1', discovery, null, store);

        assert.equal(result.valid, true);
        assert.equal(result.key_pinning.status, 'pinned');
    });

    it('should fail for invalid signature', () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const { pubPem } = makeKeyAndSign(schema);

        const discovery = {
            schema_version: '1.2',
            developer_name: 'Test Dev',
            public_key_pem: pubPem
        };
        const store = new KeyPinStore();
        const result = verifySchemaOffline(schema, 'invalid_sig', 'example.com', 'tool1', discovery, null, store);

        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.SIGNATURE_INVALID);
    });

    it('should fail for tampered schema', () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const { pubPem, sig } = makeKeyAndSign(schema);

        const tampered = { name: 'test_tool', description: 'TAMPERED' };
        const discovery = {
            schema_version: '1.2',
            developer_name: 'Test Dev',
            public_key_pem: pubPem
        };
        const store = new KeyPinStore();
        const result = verifySchemaOffline(tampered, sig, 'example.com', 'tool1', discovery, null, store);

        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.SIGNATURE_INVALID);
    });

    it('should fail for revoked key in simple list', () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const { pubPem, sig, fp } = makeKeyAndSign(schema);

        const discovery = {
            schema_version: '1.2',
            developer_name: 'Test Dev',
            public_key_pem: pubPem,
            revoked_keys: [fp]
        };
        const store = new KeyPinStore();
        const result = verifySchemaOffline(schema, sig, 'example.com', 'tool1', discovery, null, store);

        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.KEY_REVOKED);
    });

    it('should fail for revoked key in standalone doc', () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const { pubPem, sig, fp } = makeKeyAndSign(schema);

        const discovery = {
            schema_version: '1.2',
            developer_name: 'Test Dev',
            public_key_pem: pubPem
        };
        const rev = buildRevocationDocument('example.com');
        addRevokedKey(rev, fp, RevocationReason.KEY_COMPROMISE);

        const store = new KeyPinStore();
        const result = verifySchemaOffline(schema, sig, 'example.com', 'tool1', discovery, rev, store);

        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.KEY_REVOKED);
    });

    it('should reject key pin change', () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const key1 = makeKeyAndSign(schema);
        const key2 = makeKeyAndSign(schema);

        const disc1 = {
            schema_version: '1.2',
            developer_name: 'Dev',
            public_key_pem: key1.pubPem
        };
        const disc2 = {
            schema_version: '1.2',
            developer_name: 'Dev',
            public_key_pem: key2.pubPem
        };

        const store = new KeyPinStore();
        const r1 = verifySchemaOffline(schema, key1.sig, 'example.com', 'tool1', disc1, null, store);
        assert.equal(r1.valid, true);

        const r2 = verifySchemaOffline(schema, key2.sig, 'example.com', 'tool1', disc2, null, store);
        assert.equal(r2.valid, false);
        assert.equal(r2.error_code, ErrorCode.KEY_PIN_MISMATCH);
    });

    it('should fail for invalid discovery', () => {
        const store = new KeyPinStore();
        const result = verifySchemaOffline(
            { name: 'test' }, 'sig', 'example.com', 'tool1',
            { schema_version: '1.2' }, null, store
        );
        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.DISCOVERY_INVALID);
    });

    it('should fail for empty discovery', () => {
        const store = new KeyPinStore();
        const result = verifySchemaOffline(
            { name: 'test' }, 'sig', 'example.com', 'tool1',
            {}, null, store
        );
        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.DISCOVERY_INVALID);
    });
});

describe('verifySchemaWithResolver', () => {
    it('should pass with TrustBundleResolver', async () => {
        const schema = { name: 'test_tool', description: 'A test' };
        const { pubPem, sig } = makeKeyAndSign(schema);

        const wellKnown = {
            schema_version: '1.2',
            developer_name: 'Bundle Dev',
            public_key_pem: pubPem
        };
        const doc = createBundledDiscovery('example.com', wellKnown);
        const bundle = createTrustBundle('2026-01-01T00:00:00Z');
        bundle.documents.push(doc);

        const resolver = new TrustBundleResolver(bundle);
        const store = new KeyPinStore();

        const result = await verifySchemaWithResolver(schema, sig, 'example.com', 'tool1', resolver, store);
        assert.equal(result.valid, true);
        assert.equal(result.developer_name, 'Bundle Dev');
    });

    it('should fail for missing domain', async () => {
        const bundle = createTrustBundle('2026-01-01T00:00:00Z');
        const resolver = new TrustBundleResolver(bundle);
        const store = new KeyPinStore();

        const result = await verifySchemaWithResolver(
            { name: 'test' }, 'sig', 'missing.com', 'tool1', resolver, store
        );
        assert.equal(result.valid, false);
        assert.equal(result.error_code, ErrorCode.DISCOVERY_FETCH_FAILED);
    });
});
